src/uct/ib/rc/verbs/rc_verbs_ep.c

```diff
@@ -385,7 +405,26 @@ ucs_status_t uct_rc_verbs_ep_flush(uct_ep_h tl_ep, unsigned flags,
 
 ucs_status_t uct_rc_verbs_ep_fence(uct_ep_h tl_ep, unsigned flags)
 {
-    uct_rc_verbs_ep_t *ep = ucs_derived_of(tl_ep, uct_rc_verbs_ep_t);
+    uct_rc_verbs_iface_t *iface = ucs_derived_of(tl_ep->iface, uct_rc_verbs_iface_t);
+    uct_rc_verbs_ep_t *ep       = ucs_derived_of(tl_ep, uct_rc_verbs_ep_t);
+    ucs_status_t status;
+
+    if (uct_rc_ep_is_fence(&iface->super, &ep->fi)) {
+        return UCS_OK;
+    }
+
+    status = uct_rc_ep_flush(&ep->super, iface->config.tx_max_wr, flags);
+    if (status == UCS_OK) {
+        UCT_TL_EP_STAT_FENCE(ucs_derived_of(tl_ep, uct_base_ep_t));
+        return UCS_OK;
+    } else if (status != UCS_ERR_NO_RESOURCE) {
+        if (uct_rc_txqp_unsignaled(&ep->super.txqp) != 0) {
+            status = uct_rc_verbs_ep_put_short(tl_ep, NULL, 0, 0, 0);
+            if (status != UCS_OK) {
```

Comment:
plz check CQ moderation, looks like deadlock is possible waiting for resources, which were consumed by non-signalled ops

Reply:
re-implemented with respect to unsignalled events